### PR TITLE
fix: wrap shutdown handlers in promises, run them in parallel, don't exit before they are settled

### PR DESCRIPTION
**Pull request by aapolkovsky**

### CODE REVIEW

File: src/nodekit.ts [hunk 1]
```diff
@@ -91,25 +91,33 @@ export class NodeKit {
             utils: this.utils,
         });
 
-        this.addShutdownHandler(() => {
-            return new Promise((resolve) => {
-                this.tracer.close?.(() => resolve());
```
Comment (aapolkovsky) [inline src/nodekit.ts:96]: Was this optional chaining function call intentional? If it was, is it jaeger client typings issue or something else?

File: src/nodekit.ts [hunk 2]
```diff
@@ -91,25 +91,33 @@ export class NodeKit {
...
 
     addShutdownHandler(handler: ShutdownHandler) {
         this.shutdownHandlers.push(handler);
     }
+
+    private setupShutdownSignals() {
+        const signals = ['SIGTERM', 'SIGINT'] as const;
+
+        const handleSignal: ShutdownHandler = (signal) => {
+            signals.forEach((off) => process.off(off, handleSignal));
```
Comment (aapolkovsky) [inline src/nodekit.ts:107]: We should remove all handlers on shutdown signal

File: src/nodekit.ts [hunk 3]
```diff
@@ -91,25 +91,33 @@ export class NodeKit {
...
+            let code = 0;
+
+            const promises = this.shutdownHandlers.map((handler) =>
+                new Promise((resolve) => resolve(handler(signal))).catch((error) => {
+                    code = 1;
+                    this.ctx.logError('Error executing shutdown handler', error);
+                }),
+            );
+
+            Promise.allSettled(promises).then(() => process.exit(code));
```
Comment (aapolkovsky) [inline src/nodekit.ts:118]: All handlers must be resolved before program exit

File: src/nodekit.ts (hunk 2 above)
Comment (resure) [reply src/nodekit.ts:107]: Why they should be removed on shutdown signal?

Comment (resure) [reply src/nodekit.ts:107]:
```suggestion
            signals.forEach((signalName) => process.off(signalName, handleSignal));
```

File: src/nodekit.ts [hunk 4]
```diff
@@ -91,25 +91,33 @@ export class NodeKit {
...
+        };
+
+        signals.forEach((signal) => process.on(signal, handleSignal));
```
Comment (resure) [inline src/nodekit.ts:121]: Does it pass signal name to the handler?

File: src/nodekit.ts [hunk 5]
```diff
@@ -91,25 +91,33 @@ export class NodeKit {
...
+    private setupShutdownSignals() {
+        const signals = ['SIGTERM', 'SIGINT'] as const;
+
+        const handleSignal: ShutdownHandler = (signal) => {
+            signals.forEach((off) => process.off(off, handleSignal));
+
+            let code = 0;
+
+            const promises = this.shutdownHandlers.map((handler) =>
+                new Promise((resolve) => resolve(handler(signal))).catch((error) => {
```
Comment (resure) [inline src/nodekit.ts:112]: This line is little hard to read

File: src/nodekit.ts (hunk 1 above)
Comment (aapolkovsky) [reply src/nodekit.ts:96]: Now ts doesn't see any problem here. I will remove this.

File: src/nodekit.ts (hunk 4 above)
Comment (aapolkovsky) [reply src/nodekit.ts:121]: It is not properly typed but yes, signal name would be passed.

https://nodejs.org/api/process.html#signal-events

File: src/nodekit.ts (hunk 2 above)
Comment (aapolkovsky) [reply src/nodekit.ts:107]: > What will happen if they won't be removed?

In such case signals may be fired 2 and more times.

**Review by resure — APPROVED** [review]

**Pull request merged**